### DISCUSSION TITLE
Best method for creating clickable images?

### DISCUSSION THREAD

**Discussion by White-Rabbit-Scientific** [beeware/toga#1668] [Q&A]
Options would seem to be:

1. Use a button (clickable, but cannot use images)
2. Use an image (image, but not clickable)
3. Use a canvas (many options for drawing, but no image loading option)

**Comment by mhsmith** [comment]
The title was "Best method for creating clickable buttons?", but I guess you're actually asking to to make clickable *images*. In which case, see the previous discussion in #1343.

**Reply by White-Rabbit-Scientific** [reply]
Mmmm, seems we need to be careful of terminology here. I'm not sure I am clear on that.

- Picture = something like a PNG, potentially large and colorful. These are "images" in Toga.
- Icon = small, typically single color. Play, pause, stop are icons. These could be made from "canvas" objects in Toga.
- Unicode has some icon-esque options and can be used on rectangular buttons.

I read thru the previous discussion, and certainly the unicode option would not work for the stuff I am working on.

I'm keen on some simple rounded triangles (arrows) that match an existing product. I also need to add some simple graphics and other buttons with custom icons(?) on them.

Probably the canvas option would be best for me at this stage. on_press and on_release could be used for changing shading a bit when the button is pressed.

Does that sound sensible? :-)

**Reply by freakboy3742** [reply]
In short - no, it doesn't. Canvas isn't the right approach for building a button. A Button is.

"Matching an existing product" is the red flag from the perspective of Toga's design. Toga adheres to the *platform*'s look and feel. It doesn't actively support use cases that involve re-inventing a HIG from scratch, or building UIs that look identical on every platform. That means macOS buttons look like macOS buttons, Windows buttons look like Windows buttons, and so on - and as a result, macOS apps look like macOS apps and Windows apps look like Windows apps. If you've got a designer saying "I want my buttons to look like <this>", and they're talking about the entire appearance of the button, not just the icon *on* the button, Toga isn't going to be a good match.

**Reply by White-Rabbit-Scientific** [reply]
Ok, you have made it clear that Toga is designed to implement native UIs and follow each platform's HIGs :-)

Can I then ask for some really simple advice: how would you recommend implementing things like the following from Apple's Monterey OS?

<img width="388" alt="image" src="https://user-images.githubusercontent.com/6218191/200445983-fe120ba3-0b66-4291-a567-0c0dc1c663ed.png">

<img width="297" alt="image" src="https://user-images.githubusercontent.com/6218191/200446094-d2675b31-1396-4591-a9cd-888c6684b9cc.png">

<img width="71" alt="image" src="https://user-images.githubusercontent.com/6218191/200446513-90ab28fe-cd9d-445a-9a99-07926f483584.png">

Or GTK 4:

<img width="155" alt="image" src="https://user-images.githubusercontent.com/6218191/200446834-c34dd126-663b-4432-9dd9-6404e85ca2d2.png">

The learning curve (and thus time commitment) for those of us who are learning Toga and Python is steep, so suggesting that I modify Toga's source is not really as simple as it sounds. The Pycon presentation pointed to in the help docs was helpful but still it seems that coding in Objective C (and the Windows/Linux equivalents) would be required.

**Reply by freakboy3742** [reply]
How do you do this *right now*? If you just want to *use* the toolkit as currently provided, the short answer is "you don't". 

Or - you change your expectations. The first of those examples is a toga.Slider and some labels; the second is 2 text labels, a slider, and some buttons where you use Unicode glyphs (like ▶️ and 🔈) as the text label. It won't *look* the same, but it will expose the functionality so that there's a button you can push that will activate the desired functionality.

The current release of Toga is v0.3.0.dev39. It's marked as an alpha in packaging metadata. I don't know how many other ways I can flag that Toga a work in progress. There are *many* features we still want to add. Allowing icons on buttons is one of those features. 

As for contributing - I didn't say it would be easy - I said it would be *easier than trying to make Canvas behave like a button*. 

If you're still learning Python, then contributing to the Toga codebase is probably a tall ask; but then, so is trying to make a Canvas behave like another widget. However, FWIW, no Objective C coding is required to contribute to Toga. There isn't a single line of Objective C in Toga, in either the iOS or macOS backends. Toga *does* make extensive use of Objective C APIs, using a Python interface. You need to be able to *read* Objective C - or, at least, be able to read the ObjC API documentation to work out what calls can be made in Python.

**Reply by White-Rabbit-Scientific** [reply]
Thanks again. I get the feeling that I am being unhelpful. If so, just say so and I will leave you alone :-)

I'm looking at a long term commitment to a Python GUI and I think yours is the best (most sensible) option out there (even as an Alpha it is pretty good), and will only get better. So I hope I can invest some time in helping.

The info above is really helpful, so thanks !
